<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      Simpson Lab Blog &middot; SimpsonLab blog
    
  </title>

  <!-- CSS -->
  <link rel="stylesheet" href="/public/css/poole.css">
  <link rel="stylesheet" href="/public/css/syntax.css">
  <link rel="stylesheet" href="/public/css/hyde.css">
  <link rel="stylesheet" href="http://fonts.googleapis.com/css?family=PT+Sans:400,400italic,700|Abril+Fatface">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">

  <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
  </script>
</head>


  <body>

    <div class="sidebar">
  <div class="container sidebar-sticky">
    <div class="sidebar-about">
      <h1>
        <a href="/">
          Simpson Lab Blog
        </a>
      </h1>
    </div>

    <nav class="sidebar-nav">
      <a class="sidebar-nav-item active" href="/">Home</a>

      

      
      
        
          
        
      
        
          
            <a class="sidebar-nav-item" href="/about/">About</a>
          
        
      
        
          
            <a class="sidebar-nav-item" href="/archive/">Archive</a>
          
        
      
        
      
        
          
        
      
        
          
        
      

    </nav>

    <p>&copy; 2015. All rights reserved.</p>
  </div>
</div>


    <div class="content container">
      <div class="posts">
  
    
      <div class="post">
        <h1 class="post-title">
          <a href="/2015/04/08/eventalign/">
            Aligning Nanopore Events to a Reference
          </a>
        </h1>

        <span class="post-date">08 Apr 2015</span>

        <h2 id="introduction">Introduction</h2>

<p>This post describes a new module I added to our <a href="https://github.com/jts/nanopolish">nanopolish</a> software package that aligns the signal data emitted by a nanopore to a reference genome. This is in contrast to most approaches which align two DNA sequences to each other (for example a base-called read and a reference genome). To make sense of what aligning signal data to a reference genome means, I will describe at a high level my model of how nanopore sequencing works. For a more detailed and technical description, see the supplement of our <a href="http://biorxiv.org/content/early/2015/03/11/015552">preprint</a>.</p>

<h2 id="nanopore-sequencing">Nanopore Sequencing</h2>

<p>A nanopore sequencer threads a single strand of DNA through a pore embedded in a membrane. The pore allows electric current to flow from one side of the membrane to the other. As DNA transits this pore it partially blocks the flow of current, which is measured by the instrument. In Oxford Nanopore’s MinION system the measured current depends on the 5-mer that resides in the pore when the measurements are taken.</p>

<p>The MinION samples the current thousands of times per second; as 5-mers slide though the pore they should be observed in multiple samples. The MinION’s event detection software processes these samples and tries to detect points where the current level changes. These jumps indicate a new 5-mer resides in the pore. To help illustrate this I’ve reproduced a figure from our preprint below:</p>

<p><img src="/assets/simulation.svg" alt="simulation" /></p>

<p>This is a simulation from an idealized nanopore sequencing process. The black dots represent the sampled current and the red lines indicate contiguous segments that make up the detected <em>events</em>. For example the mean current was around 60 picoamps, plus a bit of noise, for the first 0.5s. The current then dropped to 40 pA for 0.1s before jumping to 52 pA and so on.</p>

<p>The event detection software writes the events to an HDF5 file. The raw kHz samples are typically not stored as the output files would be impractically large. Here’s the table of events for this simulation:</p>

<table>
  <thead>
    <tr>
      <th style="text-align: center">event index</th>
      <th style="text-align: center">mean (pA)</th>
      <th style="text-align: center">length (s)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">1</td>
      <td style="text-align: center">60.3</td>
      <td style="text-align: center">0.521</td>
    </tr>
    <tr>
      <td style="text-align: center">2</td>
      <td style="text-align: center">40.6</td>
      <td style="text-align: center">0.112</td>
    </tr>
    <tr>
      <td style="text-align: center">3</td>
      <td style="text-align: center">52.2</td>
      <td style="text-align: center">0.356</td>
    </tr>
    <tr>
      <td style="text-align: center">4</td>
      <td style="text-align: center">54.1</td>
      <td style="text-align: center">0.051</td>
    </tr>
    <tr>
      <td style="text-align: center">5</td>
      <td style="text-align: center">61.5</td>
      <td style="text-align: center">0.291</td>
    </tr>
    <tr>
      <td style="text-align: center">6</td>
      <td style="text-align: center">72.7</td>
      <td style="text-align: center">0.015</td>
    </tr>
    <tr>
      <td style="text-align: center">7</td>
      <td style="text-align: center">49.4</td>
      <td style="text-align: center">0.141</td>
    </tr>
  </tbody>
</table>

<p>To help translate events into a DNA sequence, Oxford Nanopore provides a <em>pore model</em> which describes the expected current signal for each 5-mer. The pore model is a set of 1024 normal distributions - an example might look like this:</p>

<table>
  <thead>
    <tr>
      <th style="text-align: center">5-mer</th>
      <th style="text-align: center"><script type="math/tex">\mu_k</script></th>
      <th style="text-align: center"><script type="math/tex">\sigma_k</script></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">AAAAA</td>
      <td style="text-align: center">53.5</td>
      <td style="text-align: center">1.3</td>
    </tr>
    <tr>
      <td style="text-align: center">AAAAC</td>
      <td style="text-align: center">54.2</td>
      <td style="text-align: center">0.9</td>
    </tr>
    <tr>
      <td style="text-align: center">…</td>
      <td style="text-align: center">…</td>
      <td style="text-align: center">…</td>
    </tr>
    <tr>
      <td style="text-align: center">TTTTG</td>
      <td style="text-align: center">65.3</td>
      <td style="text-align: center">1.8</td>
    </tr>
    <tr>
      <td style="text-align: center">TTTTT</td>
      <td style="text-align: center">67.1</td>
      <td style="text-align: center">1.4</td>
    </tr>
  </tbody>
</table>

<p>This indicates that the measured current is expected to be drawn from <script type="math/tex">\mathcal{N}(53.5, 1.3^2)</script> when AAAAA is in the pore and so on.</p>

<h2 id="inferring-bases-from-events">Inferring Bases from Events</h2>

<p>Using the pore model and the observed data we can solve a number of inference problems. For example we can infer the sequence of nucleotides that passed through the pore. This is the base calling problem. We can also infer the sequence of the genome given a set of overlapping reads. This is the consensus problem, which we addressed in our paper.</p>

<p>These inference problems are complicated by two important factors. First, the normal distributions for 5-mers overlap. There are 1024 different 5-mers but the signals typically range from about 40-70 pA. This can make it difficult to infer which 5-mer generated a particular event. This is partially mitigated by the structure of the data; a solution must respect the overlap between 5-mers so a position that is difficult to resolve may become clear when we look at subsequent events. Second, event detection is performed in real time and inevitably makes errors. Some events may not be detected if they are too short or if the signals for adjacent 5-mers of the DNA strand are very similar. The extreme case for the latter situation occurs when sequencing through long homopolymers - here we do not expect a detectable change in current. The opposite problem occurs as well. The event detector may split what should be a single event into multiple events due to noise in the system that looks like a change in current. Handling these artefacts is key to accurately inferring the DNA sequence that generated the events.</p>

<h2 id="aligning-events-to-a-reference">Aligning Events to a Reference</h2>

<p>The hidden Markov model we designed for the consensus problem had 5-mers of an arbitrary sequence as the backbone of the HMM, with additional states and transitions to handle the skipping/splitting artefacts. In our preprint we used this HMM to calculate a consensus sequence from a set of reads. If we make a reference genome the backbone of the HMM, we can use it to align events to the reference.</p>

<p>The new <code>eventalign</code> module of <code>nanopolish</code> exposes this functionality as a command line tool.  This program takes in a set of nanopore reads aligned in base-space to a reference sequence (or draft genome assembly) and re-aligns the reads in event space.</p>

<p>The pipeline uses <code>bwa mem</code> alignments as a guide. We start with a normal bwa workflow:</p>

<pre><code>bwa mem -x ont2d -t 8 ecoli_k12.fasta reads.fa | samtools view -Sb - &gt; alignments.bam
samtools sort alignments.bam alignments.sorted
samtools index alignments.sorted.bam
</code></pre>

<p>We can then realign in event space                                                   using nanopolish:</p>

<pre><code>nanopolish eventalign -r reads.fa -b alignments.sorted.bam -g ecoli_k12.fasta "gi|556503834|ref|NC_000913.3|:10000-20000" &gt; eventalign.tsv
</code></pre>

<p>The output looks like this:</p>

<pre><code>contig                         position  reference_kmer  read_index  strand  event_index  event_level_mean  event_length  model_kmer  model_mean  model_stdv
gi|556503834|ref|NC_000913.3|  10000     ATTGC           1           c       27470        50.57             0.022         ATTGC       50.58       1.02
gi|556503834|ref|NC_000913.3|  10001     TTGCG           1           c       27471        52.31             0.023         TTGCG       51.68       0.73
gi|556503834|ref|NC_000913.3|  10001     TTGCG           1           c       27472        53.05             0.056         TTGCG       51.68       0.73
gi|556503834|ref|NC_000913.3|  10001     TTGCG           1           c       27473        54.56             0.011         TTGCG       51.68       0.73
gi|556503834|ref|NC_000913.3|  10002     TGCGC           1           c       27474        65.56             0.012         TGCGC       66.96       2.91
gi|556503834|ref|NC_000913.3|  10002     TGCGC           1           c       27475        69.97             0.071         TGCGC       66.96       2.91
gi|556503834|ref|NC_000913.3|  10003     GCGCT           1           c       27476        67.11             0.017         GCGCT       68.08       2.20
gi|556503834|ref|NC_000913.3|  10004     CGCTG           1           c       27477        69.47             0.052         CGCTG       69.84       1.89
</code></pre>

<p>This is the complement strand (c) of a 2D Nanopore read from Nick’s <a href="http://www.gigasciencejournal.com/content/3/1/22">E. coli data</a> aligned to E. coli K12. The first event listed (event 27470) had a measured current level of 50.57 pA. It aligns to the reference 5-mer ATTGC at position 10,000 of the reference genome. The pore model indicates that events measured for 5-mer ATTGC should come from <script type="math/tex">\mathcal{N}(50.58, 1.02^2)</script>, which matches the observed data very well. The next 3 events (27471, 27472, 27473) are all aligned to the same reference 5-mer (TTGCG) indicating that the event detector erroneously called 3 events where only one should have been emitted. Note that the current for these 3 events are all plausibly drawn from the expected distribution <script type="math/tex">\mathcal{N}(51.68, 0.73^2)</script>.</p>

<p>This output has one row for every event. If a reference 5-mer was skipped, there will be a gap in the output where no signal was observed:</p>

<pre><code>gi|556503834|ref|NC_000913.3|   10009   GCACC   1       c       27489   67.52   0.028   GCACC   66.83   2.46
gi|556503834|ref|NC_000913.3|   10011   ACCGC   1       c       27490   65.17   0.012   ACCGC   65.03   1.92
</code></pre>

<p>Here we did not observe an event for the 5-mer at position 10010.</p>

<p>This module will hopefully make it easier to work with signal-level nanopore data, and help the development of improved models. The <code>eventalign</code> module can be found in the latest version of <a href="https://github.com/jts/nanopolish">nanopolish</a>.</p>

      </div>
    
  
    
      <div class="post">
        <h1 class="post-title">
          <a href="/2015/03/30/optimizing-hmm/">
            Nanopolish v0.2.0
          </a>
        </h1>

        <span class="post-date">30 Mar 2015</span>

        <p>This post describes changes I have made to <a href="https://github.com/jts/nanopolish">nanopolish</a>, our HMM-based consensus caller for Oxford Nanopore data. This post can be thought of as a long changelog with background and rationale.</p>

<h2 id="background-and-history">Background and History</h2>

<p><a href="http://nickloman.github.io/">Nick Loman</a>, Josh Quick and I started working on nanpore assembly at a hackathon at the Newton Institute in Cambridge. Our initial goal was pretty simple; we wanted to see if we could run <a href="https://github.com/thegenemyers/DALIGNER">DALIGNER</a> on nanopore data and devise a way to error correct the reads. After a lot of tinkering and “proper bioinformatics” as Nick put it (converting file formats) we were able to run <a href="http://sourceforge.net/projects/poamsa/">poa</a> on the overlapping reads that DALIGNER found. Taking poa’s consensus sequence as the error corrected read improved identity to around 92-93%. Nick was able to get <a href="http://wgs-assembler.sourceforge.net/wiki/index.php?title=Main_Page">Celera Assembler</a> running on the corrected reads and our assembly became progressively better as Nick and Josh added more data.</p>

<p>Once Nick got a single-contig assembly out of Celera Assembler we turned our attention to improving the accuracy of the final assembly. The consensus sequence that Celera Assembler called off the corrected reads had accuracy of about 98%. We knew that by working with the base-called reads, rather than the raw signal data emitted by the nanopore, we were losing a lot of information. During the winter holidays I started to write code that would use the raw current signal to call a new consensus. My initial exploratory code was in Python as the <a href="https://github.com/arq5x/poretools">poretools</a> package gave convenient access to the raw signal data encoded in ONT’s FAST5 files. I wrote a quick hidden Markov model in Python to calculate the probability of observing a sequence of nanopore signals given an arbitrary sequence. I immediately realized my Python HMM would be far too slow to run on even a bacterial genome so I decided the core algorithms would have to be written in C or C++.</p>

<p>I asked on twitter the best way to call out to a C++ library from Python and received many helpful replies (h/t to Titus Brown, Michael Crusoe and others). I settled on using <a href="https://docs.python.org/2/library/ctypes.html">ctypes</a> as this bridge between the Python frontend/poretools and the HMM in C++. I was surprised at how easy ctypes makes this - I had Python talking to a prototype C++ library in under an hour. This hybrid Python/C++ solution was just fast enough to make model development and testing possible. We spent the next month or so revising the probabilistic model of the data, developing algorithms to propose candidate consensus sequences and testing them on our E. coli data. Once the model settled we ran it on the single-contig assembly which took a few days running in parallel on Nick’s server. We wrote up a <a href="http://biorxiv.org/content/early/2015/03/11/015552">preprint</a> describing this work and posted it on BioRxiv.</p>

<h2 id="improving-the-design">Improving the design</h2>

<p>I was not satisfied with the Python/C++ hybrid design. I am sensitive to installation issues when releasing software as I have found that installing dependencies is a major source of problems for the user (although great projects like <a href="https://github.com/Homebrew/homebrew-science">homebrew-science</a> are helping a lot here). I admire Heng Li’s software where one usually just needs to run <code>git clone</code> and <code>make</code> to build the program. The initial version of nanopolish was far from this ideal as it depended on eight Python libraries that needed to be installed with <code>pip</code>. When moving between my local development version and Nick’s server I realized that installing these dependencies often failed. With this in mind I decided to rewrite the Python frontend in C++. To do this, I crucially needed a replacement for poretools which I used to access the raw data. Matei David in my group volunteered to help and wrote an excellent, intuitive C++ <a href="https://github.com/mateidavid/fast5">library</a> for parsing ONT’s FAST5 files.</p>

<p>There were additional benefits to this rewrite. In the Python version I again used poa to compute an initial multiple alignment and used this to seed the HMM realignment. In the C++ version I discarded this step, removing another dependency, by calculating the initial alignment directly from the BAM file using <a href="https://github.com/samtools/htslib">htslib</a>. This simplification, along with the much faster parsing of the FAST5 files provided by Matei’s library, reduced startup time from a few minutes to a few seconds. This has helped me iterate on the code much faster during development.</p>

<h2 id="improving-hmm-efficiency">Improving HMM efficiency</h2>

<p>Despite writing the HMM in C++ the first version of nanopolish was very slow. After the Python to C++ rewrite I focused on improving run time. During development I use a lightweight header-only <a href="https://github.com/jts/sga/blob/master/src/Util/Profiler.h">profiler</a> to keep track of where time is being spent in my program. As expected over 90% of the time was spent running the forward algorithm on the hidden Markov model. I used the amazing <a href="https://perf.wiki.kernel.org/index.php/Tutorial">perf</a> kernel profiler to explore this further. <code>perf</code> indicated that most time was in the <script type="math/tex">\log()</script> and <script type="math/tex">\exp()</script> functions. The forward algorithm on HMMs requires summing log-transformed probabilities. The naive way, <script type="math/tex">c = \log(\exp(a) + \exp(b))</script>, requires two calls to <script type="math/tex">\exp</script> and one call to <script type="math/tex">\log</script> for every state/emission pair in the HMM. This is very slow and the subject of an entire section in the classic Biological Sequence Analysis text. I remembered reading Sean Eddy’s <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002195">paper</a> on accelerating HMMER3. In this paper Sean describes how the calculation can be improved by using the transformation <script type="math/tex">c = a + \log(1 + \exp(b - a))</script> where <script type="math/tex">a \geq b</script>. On the surface this would only save a single call to <script type="math/tex">\exp</script> but Sean goes further by using a table indexed by <script type="math/tex">b - a</script> to cache <script type="math/tex">\log(1 + \exp(b - a))</script>. This method completely removes the <script type="math/tex">\exp</script> and <script type="math/tex">\log</script> calls in the inner loop of our HMM. After plugging Sean’s implementation into nanopolish we immediately had an 8-fold improvement in speed. Thanks to Sean for allowing us to use this as public domain code.</p>

<h2 id="improving-memory-layout">Improving memory layout</h2>

<p>The <code>perf</code> output also indicated that a lot of CPU time was wasted waiting on memory access. To improve cache usage and reduce the amount of data that is transferred over the memory bus, I reduced the precision of the floating point values from 64 bits to 32 bits. At the same time, I changed the memory layout of the nanopore event signals so that data accessed together was located in contiguous memory locations. This change simply interleaved two arrays, one storing event currents and one storing event durations, into a single array of structs. Finally, I pre-computed all of the scalings and transformations that need to be applied to the nanopore events (for example the current signal drifts over time and this needs to be corrected for) to reduce the work in the inner loop of the HMM.</p>

<h2 id="summary">Summary</h2>

<p>This collection of changes reduced the average time spent in the forward algorithm of the HMM from <script type="math/tex">3,000 {\mu}s</script> per call to <script type="math/tex">278 \mu s</script> per call for 100 input events and a 100bp sequence, an improvment of over 10x. This did not require any changes at the algorithm level, only minor code optimizations. My next goal is to make algorithmic improvements, primarily avoiding testing unlikely candidate sequences in the HMM. Version 0.2.0 of nanopolish is <a href="https://github.com/jts/nanopolish">here</a>.</p>

      </div>
    
  
</div>

<div class="pagination">
  
    <span class="pagination-item older">Older</span>
  
  
    <span class="pagination-item newer">Newer</span>
  
</div>

      
      


    </div>
  

  </body>
  

  <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-61162262-1', 'auto');
  ga('send', 'pageview');

</script>


</html>
