---
layout: post
title: Revising Nanopolish
draft: true
---

Last month Nick Loman, Josh Quick and I posted a [preprint](http://biorxiv.org/content/early/2015/03/11/015552) on assembling E. coli using Oxford Nanopore data. One of the challenges of working with a new sequencing platform is understanding and handling the error model. We spent a lot of time working on algorithms that use the raw signal data emitted by the nanopore to compute the final consensus sequence for the assembly. Our approach is implemented in a software package called [nanopolish](https://github.com/jts/nanopolish). At the core of this software is a hidden Markov model that calculates the probability of observing a sequence of nanopore _events_ given a proposed sequence. I'm not going to discuss the details of the probabilistic model in this post - this can be found in the supplement of our preprint - rather I'm going to discuss changes and optimizations to our implementation.

I started to work on signal-level algorithms for Oxford Nanopore data during the 2014 winter holidays. My initial exploratory code was in Python as the [poretools](https://github.com/arq5x/poretools) package gave convenient access to the raw signal data encoded in ONT's FAST5 files. My Python explorations convinced me that there was enough information in the raw signal to substantially improve the quality of our draft assembly so I set off to write a consensus algorithm. After very briefly writing a Python implementation of a PairHMM, I rewrote the core algorithms in C++ using the [ctypes](https://docs.python.org/2/library/ctypes.html) library as a bridge between the Python frontend/poretools and the HMM in C++ (h/t to Titus Brown, Michael Crusoe and others on twitter for suggesting this). This version was fast enough that we could easily test new models and ideas but very little effort was spent on performance. Once we settled on a model it took a few days to compute the final consensus using Nick's server.

After submitting the paper I spent time improving the code. My first goal was to make nanopolish easier to install, develop and run. Using Python worked well during prototyping - I could easily try out new ideas without writing much code - but as the program matured it became cumbersome to maintain the C++/Python bridge. Startup time was a problem as well; parsing the FAST5 files for 20 reads could take a few minutes to complete. This led to long write/compile/test cycles which frustrated me. Finally, the Python code depended on eight libraries. ```pip install``` very rarely worked for me without having to debug some package installation error. To fix these annoying issues I decided to rewrite all of the Python code in C++. Matei David provided a very nice C++ [library](https://github.com/mateidavid/fast5) to read the ONT FAST5 files. To read BAM files I used [htslib](https://github.com/samtools/htslib) for the first time. 

My second goal was to improve the efficiency of the hidden Markov model calculations.  I used the [perf](https://perf.wiki.kernel.org/index.php/Tutorial) kernel profiler to check for hotspots in the code. As I expected, most of the time was spent in the $$\log()$$ and $$\exp()$$ functions. When using the [Forward Algorithm](http://en.wikipedia.org/wiki/Forward_algorithm) on a hidden Markov model, one constantly needs to sum log-transformed probabilities. The naive way, $$c = \log(\exp(a) + \exp(b))$$, requires two calls to $$\exp$$ and one call to $$\log$$ for every state/emission pair in the HMM. This is very slow and the subject of an entire section in the classic Biological Sequence Analysis text. I remembered reading Sean Eddy's [paper](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002195) on accelerating HMMER3. In this paper Sean describes how the calculation can be improved by using the transformation $$c = a + \log(1 + \exp(b - a))$$ where $$a \geq b$$. On the surface this would only save a single call to $$\exp()$$ but Sean goes further by using a table indexed by $$b - a$$ to cache $$\log(1 + \exp(b - a))$$. This method completely removes the $$\exp()$$ and $$\log()$$ calls in the inner loop of our HMM. After plugging Sean's implementation into nanopolish we immediately had an 8-fold improvement in speed. Thanks to Sean for allowing us to use this as public domain code.

Next I reduced the precision of the floating point values from 64-bits to 32-bits to reduce cache usage and how much data needed to be transferred over the bus. I also reorganized the data layout in memory to reduce the number of random accesses. Finally, I pre-computed all of the scalings and transformations that need to be applied to the nanopore events (for example the current signal drifts over time and this needs to be corrected for) to reduce the work in the inner loop of the HMM. Altogether, these changes reduce run time by a factor of 12 over the initial version. There is a room for further improvement at the algorithm level, mainly how candidate consensus sequences are proposed, but those changes will have to wait. The latest version is now on github here (todo).
